DirectProgramming/DPC++FPGA/ReferenceDesigns/decompress/src/common/common.hpp

```diff
@@ -0,0 +1,398 @@
+constexpr size_t kSnappyMaxLZ77Distance = 1 << 16;
+template <size_t n>
+using SnappyLZ77InputData =
+    LZ77InputData<n, kSnappyMaxLZ77Distance, kSnappyMaxLZ77Length>;
+
+//
+// Holds an array of bytes, where valid_count indicates how many of the 'n'
+// bytes are valid. The valid bytes must be sequential and start at index 0.
+// E.g., if valid_count = 2, then byte[0] and byte[1] are valid, while byte[2],
+// byte[3], ..., byte[n-1] are not.
+//
+template <size_t num_bytes>
+struct BytePack {
+  static constexpr unsigned count_bits = fpga_tools::Log2(num_bytes) + 1;
+  static_assert(count_bits > 0);
+  static_assert(num_bytes < fpga_tools::Pow2(count_bits));
+  static constexpr size_t size = num_bytes;
+
+  unsigned char byte[num_bytes];
+  ac_uint<count_bits> valid_count;
+
+  unsigned char& operator[](int i) { return byte[i]; }
+  const unsigned char& operator[](int i) const { return byte[i]; }
+};
+
+//
+// Similar to a BytePack, but all of the bytes are valid.
+//
+template <size_t num_bytes>
+struct ByteSet {
+  static constexpr size_t size = num_bytes;
+
+  unsigned char byte[num_bytes];
+
+  unsigned char& operator[](int i) { return byte[i]; }
+  const unsigned char& operator[](int i) const { return byte[i]; }
+};
+
+//
+// returns the number of trailing zeros in an ac_int
+// E.g. 0b011101000 has 3 trailing zeros
+//
+template <int bits, bool is_signed>
+auto CTZ(const ac_int<bits, is_signed>& in) {
+  static_assert(bits > 0);
+  constexpr int out_bits = fpga_tools::Log2(bits) + 1;
+  ac_uint<out_bits> ret(bits);
+#pragma unroll
+  for (int i = bits - 1; i >= 0; i--) {
+    if (in[i]) {
+      ret = i;
+    }
+  }
+  return ret;
+}
+
+//
+// Selects a SYCL device using a string. This is typically used to select
+// the FPGA simulator device
+//
+class select_by_string : public sycl::default_selector {
+ public:
+  select_by_string(std::string s) : target_name(s) {}
+  virtual int operator()(const sycl::device& device) const {
+    std::string name = device.get_info<sycl::info::device::name>();
+    if (name.find(target_name) != std::string::npos) {
+      // The returned value represents a priority, this number is chosen to be
+      // large to ensure high priority
+      return 10000;
+    }
+    return -1;
+  }
+
+ private:
+  std::string target_name;
+};
+
+//
+// Reads 'filename' and returns an array of chars (the bytes of the file)
+//
+std::vector<unsigned char> ReadInputFile(const std::string filename) {
+  // open file stream
+  std::ifstream fin(filename);
+
+  // make sure it opened
+  if (!fin.good() || !fin.is_open()) {
+    std::cerr << "ERROR: could not open " << filename << " for reading\n";
+    std::terminate();
+  }
+
+  // read in bytes
+  std::vector<unsigned char> result;
+  char tmp;
+  while (fin.get(tmp)) {
+    result.push_back(tmp);
+  }
+  fin.close();
+
+  return result;
+}
+
+//
+// Writes the chars (bytes) from 'data' to 'filename'
+//
+void WriteOutputFile(const std::string filename,

```

Comment:
Should this be a const reference to avoid creating a temporary copy of the string?